minitask2/akhmad/behaviour.py

```diff
@@ -0,0 +1,196 @@
+
+        if not is_wall_found:
+            if all(i >= 0.5 for i in front):
+                self.publish_command(move_cmd)
+            else:
+                is_wall_found = True
+                self.publish_command(Twist())
+        else:
+            if min(right) <= 0.5 and min(front) <= 0.5:
+                self.publish_command(left_turn_cmd)
+                print("Obstacles on front and right - turning left")
+            elif min(right) <= 0.5 and min(front) >= 0.5:
+                self.publish_command(move_cmd)
+                print("No obstacles in front but on right - move forward")
+            elif min(right) >= 0.5 and min(front) <= 0.5:
+                self.publish_command(left_turn_cmd)
+                print("No obstacles on right but in front - turning left")
+            elif min(right) >= 0.5 and min(front) >= 0.5:
+                if min(right_front) <= 0.5:
+                    self.publish_command(right_turn_cmd)
+                else:
+                    self.publish_command(move_cmd)
+                print("No obstacles on right and in front - sliding right")
+
+    def obstacle_avoidance(self):
+        # continue_right and continue_left is used to
+        # keep turning in same direction as robot started
+        # otherwise it might get stuck in corners
+        if continue_right:
+            self.turn_right()
+        elif continue_left:
+            self.turn_left()
+        elif all(i >= 0.5 for i in left_front):
+            self.turn_left()
+        elif all(i >= 0.5 for i in right_front):
+            self.turn_right()
+        elif all(i >= 0.5 for i in left_back):
+            self.turn_left()
+        elif all(i >= 0.5 for i in right_back):
+            self.turn_right()
+        else:
+            self.turn_left()
+
+    def random_walk(self):
+        rospy.loginfo("Random walk")
+        global save_position, continue_right, continue_left
+        global last_pose_theta, last_pose_x, last_pose_y
+
+        continue_right = False
+        continue_left = False
+
+        # move forward 3 m and turn by random angle
+        if sqrt((self.current_pose_x - last_pose_x) ** 2 +
+                (self.current_pose_y - last_pose_y) ** 2) < 3:
+            self.publish_command(move_cmd)
+            save_position = False
+        else:
+            self.publish_command(Twist())
+            save_position = True
+            for x in range(0, random.randint(0, 120)):
+                last_pose_theta = self.current_pose_theta
+                last_pose_x = self.current_pose_x
+                last_pose_y = self.current_pose_y
+
+                self.publish_command(left_turn_cmd)
+                self.rate.sleep()
+
+    def scan_callback(self, scan):
+        global front, right_front, right_back, back, left_back, left_front, right, left
+        global last_pose_theta
+        global last_pose_x
+        global last_pose_y
+
+        front = scan.ranges[330:] + scan.ranges[0:30]
```

Comment:
I think all these scan ranges should be extracted as constants. eg `150:210` could be `SCAN_BACK_START:SCAN_BACK_END`

Reply:
This might be tedious to type. Since the limits for scan ranges are not used anywhere else, I think it is better to use them as is. But let me know if it is not a good practice

Reply:
I defo think theres a case for both approaches, but I think you might be right - extracting all of these values might be over-engineering for this task.